core/domain/exp_jobs_one_off_test.py

```diff
@@ -565,6 +609,272 @@ def test_nonhuman_committers_not_counted(self):
+
+        # Run the ExplorationContributorsSummaryOneOff job.
+        job_id = exp_jobs_one_off.ExplorationContributorsSummaryOneOffJob.create_new() # pylint: disable=line-too-long
+        exp_jobs_one_off.ExplorationContributorsSummaryOneOffJob.enqueue(job_id)
+        self.process_and_flush_pending_tasks()
+
+        self.assertEqual(
+            exp_jobs_one_off.ExplorationContributorsSummaryOneOffJob.get_output(
+                job_id),
+            []
+        )
+
+
+class OneOffExplorationFirstPublishedJobTests(test_utils.GenericTestBase):
+
+    EXP_ID = 'exp_id'
+
+    def test_first_published_time_of_exploration_that_is_unpublished(self):
+        """This tests that, if an exploration is published, unpublished, and
+        then published again, the job uses the first publication time as the
+        value for first_published_msec.
+        """
+        self.signup(self.ADMIN_EMAIL, self.ADMIN_USERNAME)
+        admin_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+        self.set_admins([self.ADMIN_USERNAME])
+
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+        owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+        owner = user_services.UserActionsInfo(owner_id)
+        admin = user_services.UserActionsInfo(admin_id)
+
+        self.save_new_valid_exploration(
+            self.EXP_ID, owner_id, end_state_name='End')
+        rights_manager.publish_exploration(owner, self.EXP_ID)
+        job_class = exp_jobs_one_off.ExplorationFirstPublishedOneOffJob
+        job_id = job_class.create_new()
+        exp_jobs_one_off.ExplorationFirstPublishedOneOffJob.enqueue(job_id)
+        self.process_and_flush_pending_tasks()
+        exploration_rights = rights_manager.get_exploration_rights(self.EXP_ID)
+
+        # Test to see whether first_published_msec was correctly updated.
+        exp_first_published = exploration_rights.first_published_msec
+        exp_rights_model = exp_models.ExplorationRightsModel.get(self.EXP_ID)
+        last_updated_time_msec = utils.get_time_in_millisecs(
+            exp_rights_model.last_updated)
+        self.assertLess(
+            exp_first_published, last_updated_time_msec)
+
+        rights_manager.unpublish_exploration(admin, self.EXP_ID)
+        rights_manager.publish_exploration(owner, self.EXP_ID)
+        job_id = job_class.create_new()
+        exp_jobs_one_off.ExplorationFirstPublishedOneOffJob.enqueue(job_id)
+        self.process_and_flush_pending_tasks()
+
+        # Test to see whether first_published_msec remains the same despite the
+        # republication.
+        exploration_rights = rights_manager.get_exploration_rights(self.EXP_ID)
+        self.assertEqual(
+            exp_first_published, exploration_rights.first_published_msec)
+
+    def test_no_action_is_performed_for_deleted_exploration(self):
+        """Test that no action is performed on deleted explorations."""
+
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
```

Comment:
Can a lot of this setup (which is duplicated across various tests) be grouped together into the setUp function? That will help I think.

This is actually a general comment. Please do this across all tests.

Reply:
Done.